src/Language/PureScript/Renamer.hs

```diff
@@ -94,54 +94,68 @@ lookupIdent name = do
-      (,) <$> renameInDecls True decls <*> traverse lookupIdent exports
+  ((moduleDecls, moduleExports), RenameState{..}) = runRename foreigns $
+    (,) <$> renameInDecls decls <*> traverse lookupIdent exports
 
 -- |
--- Renames within a list of declarations. isTopLevel is used to determine
--- whether the declarations are module members or appearing within a Let. At
--- the top level declarations are only renamed or added to the scope if they
--- are named with GenIdents (otherwise they should already have been added),
--- whereas in a Let declarations are renamed if their name shadows another in
--- the current scope. In order to prevent inner bindings earlier in the list
--- from shadowing bindings later in the list, first all of the declarations
--- are considered for renaming, and then all of the values are recursed into.
+-- Renames within a list of declarations. The list is processed in three
+-- passes:
 --
-renameInDecls :: Bool -> [Bind Ann] -> Rename [Bind Ann]
-renameInDecls isTopLevel = traverse renameDecl >=> traverse renameValuesInDecl
-  where
-  renameDecl (NonRec a name val) = updateName name <&> \name' -> NonRec a name' val
-  renameDecl (Rec ds) = Rec <$> traverse updateNames ds
+--  1) Declarations with user-provided names are added to the scope, renaming
+--     them only if necessary to prevent shadowing.
+--  2) Declarations with compiler-provided names are added to the scope,
+--     renaming them to prevent shadowing or collision with a user-provided
+--     name.
+--  3) The bodies of the declarations are processed recursively.
+--
+-- The distinction between passes 1 and 2 is critical in the top-level module
+-- scope, where declarations can be exported and named declarations must not
+-- be renamed. Below the top level, this only matters for programmers looking
+-- at the generated code or using a debugger; we want them to see the names
+-- they used as much as possible.
+--
+-- The distinction between the first two passes and pass 3 is important because
+-- a `GenIdent` can appear before its declaration in a depth-first traversal,
+-- and we need to visit the declaration first in order to rename all of its
+-- uses. Similarly, a plain `Ident` could shadow another declared in an outer
+-- scope but later in a depth-first traversal, and we need to visit the
+-- outer declaration first in order to know to rename the inner one.
+--
+renameInDecls :: [Bind Ann] -> Rename [Bind Ann]
+renameInDecls =
+      traverse (renameDecl False)
+  >=> traverse (renameDecl True)
+  >=> traverse renameValuesInDecl
 
-  renameValuesInDecl (NonRec a name val) = NonRec a name <$> renameInValue val
-  renameValuesInDecl (Rec ds) = Rec <$> traverse updateValues ds
+  where
 
-  updateName :: Ident -> Rename Ident
-  updateName name = (if isTopLevel && isPlainIdent name then pure else updateScope) name
+  renameDecl :: Bool -> Bind Ann -> Rename (Bind Ann)
+  renameDecl isSecondPass = \case
```

Comment:
Could we replace the `isSecondPass :: Boolean` arg with a simple data type so as to prevent "Boolean Blindness"? Something simple like
```haskell
data PassCount
  = FirstPass
  | SecondPass
```

Reply:
Do you really think it's worth it? `renameDecl` is a local function, only used within ten lines of its declaration; I don't think there's much of a chance for misinterpreting what that parameter is for. And it would bloat up the `if isSecondPass == isPlainIdent name` check a little.

Reply:
Mm... Probably not then. I guess mine was an overreaction.